apps/sms/js/thread_ui.js

```diff
@@ -662,50 +660,22 @@ var ThreadUI = {
     }
   },
 
-  handleForward: function thui_handleForward(forward) {
-    var request = MessageManager.getMessage(+forward.messageId);
-
-    request.onsuccess = (function() {
-      Compose.fromMessage(request.result);
-
-      Recipients.View.isFocusable = true;
-      ThreadUI.recipients.focus();
-    }).bind(this);
-
-    request.onerror = function() {
-      console.error('Error while forwarding:', this.error.name);
-    };
-  },
-
   handleActivity: function thui_handleActivity(activity) {
-    /**
-     * Choose the appropriate contact resolver:
-     *  - if we have a phone number and no contact, rely on findByAddress
-     *    to get a contact matching the number;
-     *  - if we have a contact object and no phone number, just use a dummy
-     *    source that returns the contact.
-     */
-    var findByAddress = Contacts.findByAddress.bind(Contacts);
     var number = activity.number;
-    if (activity.contact && !number) {
-      findByAddress = () => Promise.resolve(activity.contact);
-      number = activity.contact.number || activity.contact.tel[0].value;
-    }
 
     // Add recipients and fill+focus the Compose area.
     var recipientPromise;
-    if (activity.contact && number) {
-      recipientPromise = Utils.getContactDisplayInfo(findByAddress, number)
-        .then(function onData(data) {
-          data.source = 'contacts';
-          return data;
-        });
-    } else if (number) {
-      // If the activity delivered the number of an unknown recipient,
-      // create a recipient directly.
-      recipientPromise = Promise.resolve({
-        number: number,
-        source: 'manual'
+    if (number) {
+      recipientPromise = Contacts.findByAddress(number).then((contacts) => {
+        if (!contacts || !contacts.length) {
+          throw new Error('No contacts found for %s', number);
+        }
+
+        return Object.assign(
+          { source: 'contacts'}, Utils.basicContact(number, contacts)
```

Comment:
Mm, no specific reason - just randomly chosen :) But you're right, it'd be more correct if target would be object with more properties, will update